File: 0310-minimum-height-trees/0310-minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: list[list[int]]) -> list[int]:
        if n == 1:
            return [0]

        graph = {}
        degree = [0] * n
        for i in range(n):
            graph[i] = []

        for frm, to in edges:
            graph.get(frm).append(to)
            degree[frm] += 1
            graph.get(to).append(frm)
            degree[to] += 1

        leaves = []
        for i in range(n):
            if degree[i] == 1:
                leaves.append(i)

        remainingNodes = n

        while remainingNodes > 2:
            remainingNodes -= len(leaves)
            newLeaves = []

            for leaf in leaves:
                for neighbor in graph.get(leaf):
                    degree[neighbor] -= 1

                    if degree[neighbor] == 1:
                        newLeaves.append(neighbor)

            leaves = newLeaves

        return leaves

File: 0310-minimum-height-trees/0310-minimum-height-trees.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
var findMinHeightTrees = function (n, edges) {
    if (n === 1) return [0];

    const graph = new Map();
    for (let i = 0; i < n; i++) {
        graph.set(i, []);
    }

    const degree = new Array(n).fill(0);
    for (const [from, to] of edges) {
        graph.get(from).push(to);
        degree[from]++;

        graph.get(to).push(from);
        degree[to]++;
    }

    console.log(graph);
    console.log(degree);

    let leaves = [];
    for (let i = 0; i < n; i++) {
        if (degree[i] === 1) {
            leaves.push(i);
        }
    }

    let remainingNodes = n;

    while (remainingNodes > 2) {
        remainingNodes -= leaves.length;
        const newLeaves = [];

        for (let leaf of leaves) {
            for (let neighbor of graph.get(leaf)) {
                degree[neighbor]--;

                if (degree[neighbor] === 1) {
                    newLeaves.push(neighbor);
                }
            }
        }

        leaves = newLeaves;
    }

    return leaves;
};